jwst/nsclean/nsclean.py

```diff
@@ -0,0 +1,374 @@
+
+    mask : 2D boolean array
+        The mask that indicates which pixels are to be used in fitting.
+
+    Returns
+    -------
+    cleaned_image : 2D float array
+        The cleaned image.
+    """
+
+    # Instantiate the cleaner
+    cleaner = NSClean(detector, mask)
+
+    # Clean the image
+    try:
+        cleaned_image = cleaner.clean(image, buff=True)
+    except np.linalg.LinAlgError:
+        log.warning("Error cleaning image; step will be skipped")
+        return None
+
+    return cleaned_image
+
+
+def clean_subarray(detector, image, mask):
+    """Clean a subarray image.
+
+    Parameters
+    ----------
+    detector : string
+        The name of the detector from which the data originate.
+
+    image : 2D float array
+        The image to be cleaned.
+
+    mask : 2D boolean array
+        The mask that indicates which pixels are to be used in fitting.
+
+    Returns
+    -------
+    cleaned_image : 2D float array
+        The cleaned image.
+    """
+
+    # Flip the image to detector coords. NRS1 requires a transpose
+    # of the axes, while NRS2 requires a transpose and flip.
+    if detector == "NRS1":
+        image = image.transpose()
+        mask = mask.transpose()
+    else:
+        image = image.transpose()[::-1]
+        mask = mask.transpose()[::-1]
+
+    # Instantiate the cleaner
+    cleaner = NSCleanSubarray(image, mask)
+
+    # Clean the image
+    try:
+        cleaned_image = cleaner.clean()
+    except np.linalg.LinAlgError:
+        log.warning("Error cleaning image; step will be skipped")
+        return None
+
+    # Restore the cleaned image to the science frame
+    if detector == "NRS1":
+        cleaned_image = cleaned_image.transpose()
+    else:
+        cleaned_image = cleaned_image[::-1].transpose()
+
+    return cleaned_image
+
+
+def do_correction(input_model, mask_spectral_regions, n_sigma, save_mask, user_mask):
+
+    """Apply the NSClean 1/f noise correction
+
+    Parameters
+    ----------
+    input_model : data model object
+        science data to be corrected
+
+    mask_spectral_regions : boolean
+        Mask slit/slice regions defined in WCS
+
+    n_sigma : float
+        n-sigma rejection level for finding outliers
+
+    save_mask : boolean
+        switch to indicate whether the mask should be saved
+
+    user_mask : string or None
+        Path to user-supplied mask image
+
+    Returns
+    -------
+    output_model : jwst.datamodel.JwstDataModel
+        corrected data
+
+    mask_model : jwst.datamodel.JwstDataModel
+        pixel mask to be saved or None
+    """
+
+    detector = input_model.meta.instrument.detector.upper()
+    exp_type = input_model.meta.exposure.type
+    log.info(f'Input exposure type is {exp_type}, detector={detector}')
+    output_model = input_model.copy()
+
+    # Check for a user-supplied mask image. If so, use it.
+    if user_mask is not None:
+        mask_model = datamodels.open(user_mask)
+        Mask = mask_model.data.copy()
```

Comment:
User-supplied mask handling needs a couple fixes:
- The data array needs to be converted to bool to avoid an IndexError when used
- nan_pix needs to be defined to handle NaNs in the input data

Reply:
Code updates for this are in the latest commit. Let me know if it works!  Guess I should add a user-supplied mask case to the regression tests.